Update k8s to 53965 and add `verbose` for `crictl status`

miaoyq opened the pull request:
Partly fixes #166

- Update k8s to 53965
- Rename `info` to `version`
- Add `verbose` for `crictl status`

Maybe we could rename `status` to `info` .

/cc @Random-Liu @feiskyer 

Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

Random-Liu commented on cmd/crictl/version.go:
```diff
@@ -30,7 +30,7 @@ const (
 
 var runtimeVersionCommand = cli.Command{
-	Name:  "info",
+	Name:  "version",
```
I'm fine with this change.

miaoyq commented:
> If we already have crictl info, do we still need a separate crictl status. I think we don't.

Agree with you. I have rename `crictl info` to `crictl version`, so we can add a new `crictl info`,  make `status`  as a part of `info`.

miaoyq commented on cmd/crictl/status.go:
```diff
@@ -43,8 +51,8 @@ var runtimeStatusCommand = cli.Command{
 // Status sends a StatusRequest to the server, and parses the returned StatusResponse.
-func Status(client pb.RuntimeServiceClient) error {
-	request := &pb.StatusRequest{}
+func Status(client pb.RuntimeServiceClient, verbose bool) error {
```
> How do we display crictl info? Do we need to organize the information a little bit or just print json output? I feel like json is good enough for now, we could keep the yaml option.

In `map<string, string> info = 2`, the value have been in json format, so if we encode the `info` to json format, the value will not be formatted. Is there any way to make the value formatted?

feiskyer replied:
we have both version and status API in CRI, I'm ok with mapping `info` command to status API while `version` command to version API.

feiskyer replied:
> if we encode the info to json format, the value will not be formatted

what about convert `info` to yaml?

miaoyq replied:
Also not work, I make a test locally:
```go
func main() {
	test := make(map[string]interface{})
	test["Int"] = 100
	test["String"] = "string"

	jsonMap := make(map[string]interface{})
	jsonMap["Num"] = 123
	jsonMap["Str"] = "abc"
	jsonByt, _ := json.Marshal(&jsonMap)
	test["JsonStr"]	= string(jsonByt)

	byt, _ := json.Marshal(&test)
	fmt.Println("test-json:\n", string(byt))
	
	marshaledYaml, _ := yaml.Marshal(&test)
	fmt.Println("\ntest-yaml:\n",string(marshaledYaml))	

	strJson := string(byt)
	marshaledYaml, _ = yaml.Marshal(&strJson)
	fmt.Println("test-yaml-json:\n", string(marshaledYaml))	

	marshaledJson, _ := json.Marshal(&strJson)
	fmt.Println("test-json-json:\n", string(marshaledJson))	
}
```
The results are as follows：
```bash
$ go run main.go 
test-json:
 {"Int":100,"JsonStr":"{\"Num\":123,\"Str\":\"abc\"}","String":"string"}

test-yaml:
 Int: 100
JsonStr: '{"Num":123,"Str":"abc"}'
String: string

test-yaml-json:
 '{"Int":100,"JsonStr":"{\"Num\":123,\"Str\":\"abc\"}","String":"string"}'

test-json-json:
 "{\"Int\":100,\"JsonStr\":\"{\\\"Num\\\":123,\\\"Str\\\":\\\"abc\\\"}\",\"String\":\"string\"}"
```

miaoyq commented:
>  1. If we already have crictl info, do we still need a separate crictl status. I think we don't.
2. How do we display crictl info? Do we need to organize the information a little bit or just print json output? I feel like json is good enough for now, we could keep the yaml option.
3. Whether we should add the verbose flag? I feel like we should always retrieve verbose information, and display it.

@Random-Liu @feiskyer  I have addressed above comments, PTAL

Random-Liu commented on cmd/crictl/info.go:
```diff
@@ -25,36 +26,44 @@ import (
+		if err != nil {
+			return err
+		}
+		statusInfo.ExtrInfo[key] = buf
```
I guess in this way `buf` will be escaped.

miaoyq replied:
@Random-Liu  I have verified， it worked.

miaoyq replied:
This is my validation codes:
```go
package main

import "fmt"
import "encoding/json"

func main() {
	test := make(map[string]interface{})
	test["Int"] = 100
	test["String"] = "string"

	jsonMap := make(map[string]interface{})
	jsonMap["Num"] = 123
	jsonMap["Str"] = "abc"
	jsonByt, _ := json.Marshal(&jsonMap)
	test["JsonStr"]	= string(jsonByt)

	var jsonMap1 map[string]interface{}
	if err := json.Unmarshal(jsonByt, &jsonMap1); err == nil {
        	fmt.Println(jsonMap1)
	}

	test["JsonStr"] = jsonMap1
	byt, _ := json.Marshal(&test)
	fmt.Println("test-json-New:\n", string(byt))

	var testMap map[string]interface{}
	if err := json.Unmarshal(byt, &testMap); err == nil {
			fmt.Printf("%#v \n",testMap)
	}
}
```
output:
```bash
# go run main.go 
map[Num:123 Str:abc]
test-json-New:
 {"Int":100,"JsonStr":{"Num":123,"Str":"abc"},"String":"string"}
map[string]interface {}{"Int":100, "JsonStr":map[string]interface {}{"Num":123, "Str":"abc"}, "String":"string"}
```
But seems not strict, I will use your opinion. :)

Random-Liu replied:
The value in map itself could be a json.

Random-Liu commented:
@miaoyq @feiskyer Actually we just need to do something like this:
```golang
package main

import (
        "bytes"
        "encoding/json"
        "os"
)

type Status struct {
        Name  string
        Value string
}

func main() {
        status := Status{
                Name:  "name",
                Value: "value",
        }
        info := map[string]string{
                "Pid": "1234",
                "Config": `{
    "a": "b",
    "c": [
        "d",
        "e"
    ]
}`,
        }
        statusB, err := json.Marshal(&status)
        if err != nil {
                panic(err)
        }
        value := "{" + "\"Status\":" + string(statusB) + ","                                                                                       
        for k, v := range info {
                value += "\"" + k + "\"" + ":" + v + "," 
        }
        value = value[:len(value)-1]
        value += "}" 
        var out bytes.Buffer
        if err := json.Indent(&out, []byte(value), "", "  "); err != nil {
                panic(err)
        }
        out.WriteTo(os.Stdout)
}
```
```console
t$ ./test 
{
  "Status": {
    "Name": "name",
    "Value": "value"
  },
  "Pid": 1234,
  "Config": {
    "a": "b",
    "c": [
      "d",
      "e"
    ]
  }
}
```

Random-Liu commented:
And we may need to decide whether field should start with capitalized letter or not.
We should make it consistent.

miaoyq commented on cmd/crictl/version.go:
```diff
@@ -18,45 +18,43 @@ package main
+		err := Version(runtimeClient, criClientVersion)
 		if err != nil {
-			return fmt.Errorf("Getting status of runtime failed: %v", err)
+			return fmt.Errorf("Getting the runtime version failed: %v", err)
```
will do.

Random-Liu replied:
Please address this.

feiskyer commented:
> Actually we just need to do something like this:

Cool. @miaoyq could you format json in this way?

> And we may need to decide whether field should start with capitalized letter or not.

Shouldn't we keep consistent with what runtimes returns?

miaoyq commented:
@feiskyer OK, I'm working on this.

miaoyq commented:
@feiskyer @Random-Liu  Updated.

feiskyer commented:
@miaoyq Could you also format output when user chooses yaml format?

@Random-Liu Does containerd provide those extra info now? need to verity the output.

miaoyq commented:
@feiskyer OK.

miaoyq commented:
@feiskyer  Updated.

feiskyer commented on cmd/crictl/info.go:
```diff
@@ -25,36 +26,44 @@ import (
-	Usage: "Display runtime version information",
+var runtimeStatusCommand = cli.Command{
+	Name:      "info",
+	Usage:     "Display status and extra information of the container runtime",
```
nit: change usage according to Random's comments.

miaoyq replied:
Sorry for missing this.

miaoyq commented:
@feiskyer  Thanks for reminder, Done.

Random-Liu commented:
> Does containerd provide those extra info now? need to verity the output.

Not yet. We are planning to add it this release. @miaoyq Could you add some basic information for `info` in containerd? For the first step, maybe we could just return current cri-containerd config in json format.

Random-Liu commented on cmd/crictl/info.go:
```diff
@@ -17,44 +17,75 @@ limitations under the License.
+		return err
+	}
+
+	statusByte, err := json.Marshal(r.Status)
```
We can make this a helper function, because we'll need it for container/sandbox/image status.
I'm fine with doing that in another PR.

miaoyq replied:
Will do late. :)

Random-Liu commented:
LGTM other than one comment.

miaoyq commented:
@Random-Liu  Addressed.

miaoyq commented:
> @miaoyq Could you add some basic information for info in containerd? For the first step, maybe we could just return current cri-containerd config in json format.

@Random-Liu  Do you means adding some basic information for `info` in **crictl**?

Random-Liu commented:
@miaoyq I mean return those information from `Status` function of cri-containerd.

Then `crictl` will print it out automatically with your PR, right?

Random-Liu commented:
/lgtm

miaoyq commented:
@Random-Liu  Ok, got it, thanks.
